Documentation for Workspace Configs (#3788)
vercel/turbo: docs/pages/repo/docs/core-concepts/monorepos/configuring-workspaces.mdx

---
title: Configuring Workspaces
description: Turborepo lets you configure each workspace
---

import Callout from "../../../../../components/Callout";

# Configuring Workspaces

> Most monorepos can declare a `turbo.json` in the root directory with a uniform

Reviewer: This felt a little...maybe presumptuous? Not sure what word I want to use.

At any rate, I think we can jump straight to "While uniformity across your codebase tends to make things easier to handle, not all packages and workspaces are built the same. To accommodate this..."

Author: Talked about this internally and sticking with this, but open to word smithing!

set of tasks that apply to all workspaces. But sometimes, a monorepo can contain
workspaces that need to configure their tasks differently. To accommodate this,
Turborepo enables you to extend the root configuration with a `turbo.json` in
any workspace. This flexibility enables a more diverse set of apps and packages
to co-exist, and allows workspace owners to maintain specialized tasks and

> configuration without affecting other inhabitants of the monorepo.

Reviewer: "without affecting developers working in the monorepo."

Additionally, do we want to be saying "monorepo" specifically? `turbo` can be used in multirepo situations, as well.

Author: this whole page is inside the monorepos section. workspace overrides doesn't apply to single package repos

## How it Works

> To override any of the configuration for a task, add a `turbo.json` file in any

Reviewer: "any of the configuration for a task from the root" or something of that nature. I'm thinking we need to mention the root oooone more time here.

> workspace of your monorepo with a top level `extends` key:

Reviewer: top level -> top-level

```jsonc filename="turbo.json"
{
  "extends": ["//"],
  "pipeline": {

>     // custom configuration for the workspace here

Reviewer: I think we can bring slightly more color here. Something like...

// Overrides for this workspace
// dependsOn, inputs, outputs, etc.

Author: added some more comments, but refraining from adding speciifc keys, because (1) they get out of date, and (2) there isn't anything to compare specific keys to in this intro section

  }
}
```

<Callout>

>   For now, the only valid value for the `extends` key is `["//"]`. The

Reviewer: For now, the only valid value for the `extends` key is `["//"]`. The
`//` syntax refers to the "root", similar to how it's used to
[declare root tasks][1].

Author: not seeing the difference here

`//` syntax, you may recall, signifies "root", similar to how it's used to
[declare root tasks][1].
</Callout>

> Extended configuration in a workspace can reset or override any of [the

Reviewer 2: I interpret reset as a special case of overriding
```suggestion
Extended configuration in a workspace can override any of [the
```

> Extended configuration in a workspace can reset or override any of [the

Reviewer: will* reset and* override

Reviewer: A bit more intentional! I *know* what's going to happen.

Author: I meant "reset" and "override" as different things. But I've removed reset entirely based on @chris-olszewski's suggestion that "reset" is a special case of override, which I agree with.

configuration for a pipeline task][2]. If you don't include a key, it will inherit

> the value from the extended `turbo.json`.

Reviewer 3: nit: "extended" here is a little ambiguous. It took me a sec to realize it meant the root `turbo.json`. Maybe we could call it "root" or "the `turbo.json` at the root"?

Author: I chose to be ambigious because the "can only extend from root" is framed as a starting point, and eventually we want to extend from any worksapce.

## Examples

To illustrate, let's look at a couple of use cases.

### Different Frameworks

> Let's say your monorepo has a many Next.js apps, and one Sveltekit app. By

Reviewer 3: ```suggestion
Let's say your monorepo has multiple Next.js apps, and one SvelteKit app. By
```

> Let's say your monorepo has a many Next.js apps, and one Sveltekit app. By

Reviewer: Looks like an extra "a"

default, both frameworks create their build with the `build` script in their

> respective `package.json`s. With a single `turbo.json` you can configure this

Reviewer: "With a single `turbo.json`,"

comma

Reviewer: Also think we need to mention "at the root" or something here.

with:

```jsonc filename="turbo.json"
{
  "pipeline": {
    "build": {

>       "outputs": [".next/**", ".svelte-kit/**"],

Reviewer 2: Should we add another key to show off that you don't need to repeat keys that don't need overrides?
Something like basic `"inputs": ["src/**/*.js"]`

Author: this is noted i the Comparison to Workspace tasks section

    }
  }
}
```

Notice how both `.next/**` and `.svelte-kit/**` need to be specified as outputs,

> even though though the Next.js apps do not generate this output. With Workspace

Reviewer: two `though`s

> overrides, you can instead add a new configuration at

Reviewer: Called it "Worskspace Overrides" here. Hmmm, does that name feel more natural...

Reviewer: Or perhaps you were using it as a common noun here?

Author: I was indecisive while writing about "worksapce overrides" vs "workspace config" and then didn't clean up after myself. fixed now

`apps/my-svelte-kit-app/turbo.json`, and remove the config from the root
configuration:

```jsonc filename="apps/my-svelte-kit-app/turbo.json"
{
  "extends": ["//"],
  "pipeline": {
    "build": {
      "outputs": [".svelte-kit/**"]
    }
  }
}
```

```diff filename="turbo.json"
{
  "pipeline": {
    "build": {
-      "outputs": [".next/**", ".svelte-kit/**"]
+      "outputs": [".next/**"]
    }
  }
}
```

> ### Specialized Tasks

Reviewer 2: In our running tasks docs we use "specific workspace-tasks" as the term for tasks that only appear in a single workspace. Later in this doc we use "Workspace Specific Task". I think we need to unify how we talk about composable turbo.json compared to workspace#task.

Author: Agreed, but this was slightly different. I wanted to show that you can add _new_ tasks to workspaces taht don't exist anywhere else. Workspace-Specific tasks, as of today, are just about specialized configuration of _existing_ tasks.

In another example, say that the `build` task in one workspace `dependsOn` a
`compile` task. You could universally declare it as `dependsOn: ["compile"]`.

> This means that your root `turbo.json` has to have a empty `compile` task entry:

Reviewer: an*

```json filename="turbo.json"
{
  "extends": ["//"],
  "pipeline": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {}
  }
}
```

> With a Workspace override, you can move that `compile` task into the

Reviewer 3: Is the W meant to be capitalized in Workspace?

> `apps/my-custom-app/turbo.json` and remove it from the root.

Reviewer: Would love to pepper in a little bit of higher-level motivation here.

"remove it from the root...to improve the clarity of your root tasks"
"to move the definition of the task closer to where it is used"

something of that nature

Author: done

> ## Relationship to `Workspace#Task` syntax

Reviewer: I'd lowercase the `W` in `Workspace`.

Reviewer: I'm also thinking it would be more clear as "Comparing to..." instead of "Relationship to..."

> ## Relationship to `Workspace#Task` syntax

Reviewer 2: Syntax reads to me as we're using the same feature, just in a different way.
```suggestion
## Relationship to workspace-tasks defined in root `turbo.json`
```

> You may be wondering: how does this relate to [`workspace#task` entries][3] in the root turbo.json?

Reviewer: This got a little verbose. Let's see how can I do some cutting...

"At first glance, Workspace Configs may sound a lot like the `workspace#task` syntax. However, Workspace Configs let us be more more intentional about our task while writing less code. For example..."

These features are very similar, with one major difference. When you declare a specific workspace task
in the root turbo.json, it is _completely_ preferred over the baseline task configuration. With a
Workspace Override, the task configuration is merged. Consider the example of the monorepo with
four Next.js apps and a Sveltekit app again. Without a workspace override, you may configure your
root `turbo.json` like this:

```json filename="turbo.json"
{
  "pipeline": {
    "build": {
      "outputMode": "hash-only",
      "outputs": [".next/**"],
    },
    "my-sveltekit-app#build": {
      "outputs": [".svelte-kit/**"]
    }
  }
}
```

> This works, but it's important to understand that the `my-sveltekit-app#build`

Reviewer: Picking up from comment above, seeing how I can do more with less...

"It might have snuck by us that our `my-sveltekit-app#build` task lost the `outputMode` when we really were only meaning to change the cache outputs. We would need to duplicate the `outputMode` key get that configuration back.

If we have many configurations, this ends up being a lot of duplicated code to maintain. With Workspace Configs, you don't need to duplicate any configuration. Instead, you only need to override what needs to be changed."

Author: reworded to talk more about the duplication

completely ovewrites `build` for the Sveltekit app. Any other config (such as
`outputMode`) in the `build` config would be lost. To retain the `build` config,
you would have to duplicate all the config including the override. Other than

> duplication, the obvious problem with this setup is that it can be difficult to

Reviewer 3: ```suggestion
redundancy, the obvious problem with this setup is that it can be difficult to
```

understand over time which duplications are intentional and which ones are added
just because someone wanted to add a custom `outputs` for their workspace.

> With workspace overrides, you don't need duplicate any configuration, just

Reviewer: "workspace overrides" again! Hmmm, maybe it is more natural...

</Callout>

## Limitations

Although the general idea is the same, Workspace Configs come with a set of
guardrails that can prevent workspaces from creating confusing situations. These
guardrails are listed here to make it clear that they are intentional, rather
than accidental:

- Workspace Configurations cannot use [the `workspace#task` syntax][3], because
  the `workspace` is inferred based on the location of the config, and it is not
  possible to extend the configuration for another workspace.
- Workspace Configurations cannot override anything outside the `pipeline` key.
  For example, it is not possible to override `globalEnv` or
  `globalDependencies`. We expect that monorepo owners should control this
  absolutely, and if this config is not _truly_ global, it should not be
  configured that way.
- Root turbo.json cannot use the `extends` key.

If you have a use case for any of these, please [file an issue][4]!

## Troubleshooting

In large monorepos, it can sometimes be difficult to understand how Turborepo is
interpreting your configuration. To help, we've added a `resolvedTaskDefinition`
to the Dry Run output. If you run `turbo run build --dry-run`, for example, the

> output will include the combination of all `turbo.json` configuration that were

Reviewer 3: ```suggestion
output will include the combination of all `turbo.json` configurations that were
```

  }
}
```

But notice that both `.next/**` and `.svelte-kit/**` need to be specified as
outputs, even though Next.js apps do not generate a `.svelte-kit` directory, and
vice versa. With Workspace Configurations, you can instead add custom
configuration in the SvelteKit workspace in `apps/my-svelte-kit-app/turbo.json`:

```jsonc filename="apps/my-svelte-kit-app/turbo.json"
{
  "extends": ["//"],
  "pipeline": {
    "build": {
      "outputs": [".svelte-kit/**"]
    }
  }
}
```

and remove the config from the root configuration:

```diff filename="turbo.json"
{
  "pipeline": {
    "build": {
-      "outputs": [".next/**", ".svelte-kit/**"]
+      "outputs": [".next/**"]
    }
  }
}
```

This not only makes each configuration easier to read, it puts the configuration
closer to where it is used.

### Specialized Tasks

In another example, say that the `build` task in one workspace `dependsOn` a
`compile` task. You could universally declare it as `dependsOn: ["compile"]`.
This means that your root `turbo.json` has to have an empty `compile` task
entry:

```json filename="turbo.json"
{
  "pipeline": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {}
  }
}
```

> With a Workspace Configurations, you can move that `compile` task into the

Reviewer: Little bit of grammar. 

a [plural]

`apps/my-custom-app/turbo.json`,

```json filename="apps/my-app/turbo.json"
{
  "extends": ["//"],
  "pipeline": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {}
  }
}
```

and remove it from the root:

```diff filename="turbo.json"
{
  "pipeline": {
+    "build": {}
-    "build": {
-      "dependsOn": ["compile"]
-    },
-    "compile": {}
  }
}
```

Now, the owners of `my-app`, can have full ownership over their `build` task,
but continue to inherit any other tasks defined at the root.

## Comparison to workspace-tasks defined in root `turbo.json`

At first glance, Workspace Configurations may sound a lot like the

> [`workspace#task`` syntax][3] in the root `turbo.json`. The features are

Reviewer: Backticks formatting got messed up here!

similar, but have one significant difference: when you declare a workspace-task
in the root ` turbo.json``, it _completely_ overwrites the baseline task
configuration. With a Workspace Config, the task configuration is merged

> instead. Consider the example of the monorepo with multiple Next.js apps and a

Reviewer: Line break before `Consider`

Sveltekit app again. Without a workspace task, you might configure your root `turbo.json` like this:

```jsonc filename="turbo.json"
{
  "pipeline": {
    "build": {
      "outputMode": "hash-only",
      "inputs": ["src/**"],
      "outputs": [".next/**"],
    },
    "my-sveltekit-app#build": {
      "outputMode": "hash-only", // must duplicate this
      "inputs": ["src/**"], // must duplicate this
      "outputs": [".svelte-kit/**"]
    }
  }
}
```

This works, but it's important to understand that the `my-sveltekit-app#build`
completely ovewrites `build` for the Sveltekit app. Any other config (such as
`outputMode`) in the `build` config would need to be duplicated. With Workspace

> Configurations, you don't need to duplicate any configuration, just override

Reviewer: Period instead of comma. "...any configuration. Just override..."

---
title: Configuring Workspaces
description: Turborepo lets you configure each workspace
---

import Callout from "../../../../../components/Callout";

# Configuring Workspaces

> Most monorepos can declare a `turbo.json` in the root directory with a uniform
> [pipeline][2] that apply to all workspaces. Sometimes, a monorepo can contain
> workspaces that need to configure their tasks differently. To accommodate this,
> starting in version 1.8, Turborepo enables you to extend the root configuration
> with a `turbo.json` in any workspace. This flexibility enables a more diverse
> set of apps and packages to co-exist, and allows workspace owners to maintain
> specialized tasks and configuration without affecting other apps and packages of
> the monorepo.

Reviewer 4: Do we need this bit `Most monorepos can declare a `turbo.json` in the root directory with a uniform[pipeline][2] that applies to all workspaces.` ? I think its cleaner to jump straight in like this:
```suggestion
From version 1.8 you can extend the root configuration of _any_ workspace with a `turbo.json` file. This flexibility enables a more diverse
set of apps and packages to co-exist, and allows workspace owners to maintain specialized tasks and configuration without affecting other apps and packages of the monorepo.
```

## How it Works

To override the configuration for any task defined in the root `turbo.json`, add
a `turbo.json` file in any workspace of your monorepo with a top-level `extends`
key:

```jsonc filename="apps/my-app/turbo.json"
{
  "extends": ["//"],
  "pipeline": {
    "build": {
      // custom configuration for the build task in this workspace
    },
    // new tasks only available in this workspace
    "special-task": {},
  }
}
```

<Callout>
  For now, the only valid value for the `extends` key is `["//"]`.
`//` is a special name used to identify the root directory of the monorepo.
</Callout>

> Configuration in a workspace can override any of [the configuration for a
> pipeline task][2]. If you don't include a key, the configuration is inherited
> from the extended `turbo.json`.

Reviewer 4: ```suggestion
Configuration in a workspace can override any of [the configurations for a
pipeline task][2]. If you don't include a key, the configuration is inherited
from the extended `turbo.json`.
```
